File: Platforms/Common/Reign_Video_D3D11_Component/States/DepthStencilState.cpp
#include "pch.h"
#include "DepthStencilState.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	DepthStencilStateError DepthStencilStateCom::Init(VideoCom^ video, DepthStencilStateDescCom^ desc)
	{
		null();
		this->video = video;

		ID3D11DepthStencilState* stateTEMP;
		if (FAILED(video->device->CreateDepthStencilState(desc->desc, &stateTEMP)))
		{
			return DepthStencilStateError::DepthStencil;
		}
		state = stateTEMP;

		return DepthStencilStateError::None;
	}

	DepthStencilStateCom::~DepthStencilStateCom()
	{
		if (state) state->Release();
		null();
	}

	void DepthStencilStateCom::null()
	{
		state = 0;
	}
	#pragma endregion

	#pragma region Methods
	void DepthStencilStateCom::Enable()
	{
		video->deviceContext->OMSetDepthStencilState(state, 0);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/States/DepthStencilState.cs
﻿using System;
using Reign.Core;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class DepthStencilState : DisposableResource, IDepthStencilState
	{
		#region Properties
		private DepthStencilStateCom com;
		#endregion

		#region Constructors
		public DepthStencilState(IDisposableResource parent, IDepthStencilStateDesc desc)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				com = new DepthStencilStateCom();
				var error = com.Init(video.com, ((DepthStencilStateDesc)desc).com);

				if (error == DepthStencilStateError.DepthStencil) Debug.ThrowError("DepthStencil", "Failed to create DepthStencil");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Enable()
		{
			com.Enable();
		}
		#endregion
	}
}